src/lib/src/common/workspace.spec.ts

```diff
@@ -1,331 +1,442 @@
+  });
+
+  it('returns empty object if there are no markers for a given path', () => {
+    // given
+    const workspace = new Workspace();
+    workspace.setMarkerValue('some/other/path', 'anUnrelatedField', 'bar');
+
+    // when
+    const actualMarker = workspace.getMarkers('some/path');
+
+    // then
+    expect(actualMarker).toEqual({});
+  });
+
+});
+
+describe('Workspace marker polling', () => {
+  it('observes external data sources and updates marker values, accordingly', fakeAsync(() => {
+    // given
+    const observer: MarkerObserver<ElementState> = {
+      path: 'sample/path',
+      field: 'testStatus',
+      observe: () => Promise.resolve(ElementState.LastRunSuccessful),
+      stopOn: (currentState) => currentState !== ElementState.Running
+    }
+    const workspace = createWorkspaceWithRootFolder(observer.path);
+
+    // when
+    workspace.observeMarker(observer);
+    tick();
+
+    // then
+    expect(workspace.getMarkerValue(observer.path, observer.field)).toEqual(ElementState.LastRunSuccessful);
+  }));
+
+  it('keeps polling until exit condition is reached', fakeAsync(() => {
+    // given
+    const stateBefore = ElementState.Running;
+    const stateAfter = ElementState.LastRunFailed;
+    const observer: MarkerObserver<ElementState> = {
+      path: 'sample/path',
+      field: 'testStatus',
+      observe: () => Promise.resolve(invocations++ < invocationsUntilChange ? stateBefore : stateAfter),
+      stopOn: (currentState) => currentState !== stateBefore
+    }
+    const workspace = createWorkspaceWithRootFolder(observer.path);
+    const invocationsUntilChange = 3;
+    let invocations = 0;
+
+    // when
+    workspace.observeMarker(observer);
+    tick();
+
+    // then
+    expect(workspace.getMarkerValue(observer.path, observer.field)).toEqual(stateAfter);
+    expect(invocations).toEqual(invocationsUntilChange + 1);
+  }));
+
+  it('stops polling when the path disappears from the workspace', fakeAsync(() => {
+    // given
+    const stateBefore = ElementState.Running;
+    const stateAfter = ElementState.LastRunFailed;
+    const observer: MarkerObserver<ElementState> = {
+      path: 'sample/path',
+      field: 'testStatus',
+      observe: () => new Promise((resolve) => setTimeout(resolve, 1))
+        .then(() => Promise.resolve(invocations++ < invocationsUntilChange ? stateBefore : stateAfter)),
+      stopOn: (currentState) => currentState !== stateBefore
+    }
+    const workspace = createWorkspaceWithRootFolder(observer.path);
+    const invocationsUntilChange = 3;
+    const invocationsUntilReload = 2;
+    let invocations = 0;
+
+    workspace.observeMarker(observer);
+    tick(invocationsUntilReload);
+
+    // when
+    workspace.reload({name: '', path: 'unobserved/path', type: ElementType.File, children: []});
+    tick();
+
+    // then
+    expect(invocations).toEqual(invocationsUntilReload);
+    expect(workspace.getMarkerValue(observer.path, observer.field)).toEqual(stateBefore);

```

Comment:
this can actually be argued over, since this path is no longer known to the workspace, right?
Maybe a cleanup needs to be defined additionally that will remove all markers no longer known to the workspace. Doing this explicitly will still give the opportunity to check "recently lost" elements

Reply:
* added a method `clearStaleMarkers()`
* by default, it is invoked during workspace reload, but it can be skipped using an optional boolean flag
* also, found a potential bug (`pathToElement` map was not cleared out on reload, potentially causing paths themselves to become stale)
* this, in turn, revealed a wrong import in the tests of `NavigationComponent`